### PR TITLE
Reduce IO string allocations

### PR DESCRIPTION
**Pull request by nirvdrum**

### CODE REVIEW

File: src/main/java/org/truffleruby/extra/ffi/PointerNodes.java [hunk 1]
```diff
@@ -47,13 +51,37 @@
...
+                checkNullPointerNode = insert(CheckNullPointerNode.create());
+            }
+
+            checkNullPointerNode.executeCheck(ptr);
+        }
+
+    }
+
+    @NodeChildren({ @NodeChild("pointer") })
+    public abstract static class CheckNullPointerNode extends RubyNode {
```
Comment (eregon) [inline src/main/java/org/truffleruby/extra/ffi/PointerNodes.java]: This is not optimal for the footprint but probably not a big deal either.
One way to be better on footprint but still deduplicate code would be to define a static method taking a Pointer + BranchProfile.

File: src/main/ruby/core/posix.rb [hunk 2]
```diff
@@ -136,6 +136,8 @@ def self.attach_function_eagerly(native_name, argument_types, return_type,
           end
         elsif return_type == :pointer
           result = Truffle::FFI::Pointer.new(Truffle::Interop.as_pointer(result))
+        elsif return_type == :ssize_t && Truffle.invoke_primitive(:fixnum_fits_into_int, result)
```
Comment (eregon) [inline src/main/ruby/core/posix.rb]: The explicit check `fixnum_fits_into_int` is unneeded, `FixnumOperations.lower` already does it and handle if it doesn't fit.
Can you remove it?

File: src/main/ruby/core/io.rb [hunk 3]
```diff
@@ -352,6 +356,17 @@ def inspect # :nodoc:
...
+    Truffle::Boot.delay do
+      if Truffle::Boot.get_option('polyglot.stdio')
+        def fill_read(io, count)
+          buffer = Truffle::POSIX.read_string_blocking(io, count)
+          bytes_read = buffer ? buffer.bytesize : 0
+
+          [buffer, bytes_read]
+        end
+      end
+    end
```
Comment (eregon) [inline src/main/ruby/core/io.rb]: Could you move it right below `fill_read` above so the possible redefinition stays together?

File: src/main/ruby/core/io.rb [hunk 4]
```diff
@@ -352,6 +356,17 @@ def inspect # :nodoc:
       format '#<IO::InternalBuffer:0x%x total=%p start=%p used=%p data=%p %s>',
              object_id, @total, @start, @used, @storage, content
     end
+
+    Truffle::Boot.delay do
+      if Truffle::Boot.get_option('polyglot.stdio')
+        def fill_read(io, count)
```
Comment (eregon) [inline src/main/ruby/core/io.rb]: This is of course not pretty but seems a decent workaround for now.
Maybe we can make it so polyglot reads (and writes) also return a native buffer to simplify and be consistent?
It won't be optimal for polyglot IO performance but probably good enough. cc @chrisseaton

File: src/main/ruby/core/io.rb (hunk 3 above)
Comment (nirvdrum) [reply src/main/ruby/core/io.rb]: Sure. I just put it at the bottom of the class so all such redefinitions would be grouped together.

Comment (eregon) [reply src/main/ruby/core/io.rb]: That works too :)

**Comment by eregon** [comment]
Let's merge this (once CI passes)

**Pull request merged**